Trajectory message contains waypoints that are not strictly increasing in time.

### Commit
At which commit was the issue/bug observed?

### Steps to reproduce
On a real robot: Make a sequence, where the second command has a blend radius, e.g.

    r.move(Ptp(goal=Pose(position=Point(0.0, 0.0, .9), orientation=Quaternion(0,0,0,1)),
                    vel_scale=SLOW_VEL_SCALE,
                    acc_scale=ACC_SCALE))
    seq = Sequence()
    seq.append(Ptp(goal=Pose(position=Point(0.0, 0, .9), orientation=Quaternion(0,0,0,1)),
                    vel_scale=SLOW_VEL_SCALE,
                    acc_scale=ACC_SCALE))
    seq.append(Ptp(goal=Pose(position=Point(0.2, 0, .9), orientation=Quaternion(0,0,0,1)),
                    vel_scale=SLOW_VEL_SCALE,
                    acc_scale=ACC_SCALE),
		blend_radius=0.099)
    seq.append(Ptp(goal=Pose(position=Point(0.2, 0.2, .9), orientation=Quaternion(0,0,0,1)),
                   vel_scale=SLOW_VEL_SCALE,
                   acc_scale=ACC_SCALE),
               blend_radius=0.099)
    seq.append(Ptp(goal=Pose(position=Point(0, 0.2, .9), orientation=Quaternion(0,0,0,1)),
                   vel_scale=SLOW_VEL_SCALE,
                   acc_scale=ACC_SCALE))
    r.move(seq)

e.g. https://github.com/ct2034/my_pilz_sandbox/blob/627759a2e0a47ac6b08d7f21c259e429f78cf21f/my_pilz_sandbox/scripts/seq.py

### Expected behavior
Robot executes sequence

### Observed behavior
`[ERROR] [1555069916.455739612]: Trajectory message contains waypoints that are not strictly increasing in time.`
(This happens to me only on a hardware robot)

### Reference to test

- on robot:
https://github.com/PilzDE/pilz_industrial_motion/blob/fix/not_srictly_incresaing_in_time/pilz_robot_programming/test/acceptance_tests/acceptance_test_blending.py#L296

- analyzing trajectories:
https://github.com/PilzDE/pilz_industrial_motion/blob/fix/not_srictly_incresaing_in_time/pilz_trajectory_generation/test/integrationtest_command_list_manager.cpp#L207

--- a/pilz_industrial_motion_testutils/include/pilz_industrial_motion_testutils/sequence.h
+++ b/pilz_industrial_motion_testutils/include/pilz_industrial_motion_testutils/sequence.h
@@ -68,7 +68,7 @@ public:
   MotionCmd &getCmd(const size_t index_cmd);
 
   void setAllBlendRadiiToZero();
-  void setBlendRadii(const size_t index_cmd, const double blend_radius);
+  void setBlendRadius(const size_t index_cmd, const double blend_radius);
   double getBlendRadius(const size_t index_cmd) const;
 
   /**
@@ -110,7 +110,7 @@ inline double Sequence::getBlendRadius(const size_t index_cmd) const
   return cmds_.at(index_cmd).second;
 }
 
-inline void Sequence::setBlendRadii(const size_t index_cmd, const double blend_radius)
+inline void Sequence::setBlendRadius(const size_t index_cmd, const double blend_radius)
 {
   cmds_.at(index_cmd).second = blend_radius;
 }

--- a/pilz_robot_programming/test/acceptance_tests/acceptance_test_blending.py
+++ b/pilz_robot_programming/test/acceptance_tests/acceptance_test_blending.py
@@ -45,6 +45,8 @@ def start_program():
 
     _testUnreachableFirstPose(robot)
 
+    _testBlendRadiusOnSecondCommand(robot)
+
 # Tests a sequence which contains Lin motions commanded in joint space
 # instead of cartesian space.
 #
@@ -275,6 +277,38 @@ def _testLinLinBlendWithOriChange(robot):
         return
     _linLinBlend(robot, 1.57)
     _askSuccess(_testLinLinBlendWithOriChange.__name__)
+
+# Tests the blending of three motions where only the second has a radius.
+# 
+# Test Sequence:
+#   1. Move to start position.
+#   2. Execute two blended motions.
+#
+#
+# Expected Results:
+#   1. Robot moves to start position.
+#   2. Robot executes a trajectory which consists of two blended motions.
+#       a. Robot reached goal pose of third trajectory.
+#       b. Robot performed continuous motion (without stop).
+#       c. Robot performed continuous orientation change (continuous rotational velocity).
+#       d. Robot makes smooth curve between into last trajectory.
+#
+def _testBlendRadiusOnSecondCommand(robot):
+    if (_askPermission(_testBlendRadiusOnSecondCommand.__name__) == 0):
+        return
+
+    robot.move(Ptp(goal=Pose(position=Point(0.0, 0.0, 0.9), orientation=from_euler(0,0,0)), 
+                   vel_scale=DEFAULT_PTP_VEL))
+    seq1 = Sequence()
+    seq1.append(Ptp(goal=Pose(position=Point(0.2, 0.0, 0.9), orientation=from_euler(0,0,0)), 
+                    vel_scale=DEFAULT_PTP_VEL, acc_scale=DEFAULT_PTP_ACC))
+    seq1.append(Ptp(goal=Pose(position=Point(0.2, 0.2, 0.9), orientation=from_euler(0,0,0)), 
+                    vel_scale=DEFAULT_PTP_VEL, acc_scale=DEFAULT_PTP_ACC), blend_radius=.1)
+    seq1.append(Ptp(goal=Pose(position=Point(0.0, 0.2, 0.9), orientation=from_euler(0,0,0)), 
+                    vel_scale=DEFAULT_PTP_VEL, acc_scale=DEFAULT_PTP_ACC))
+    robot.move(seq1)
+
+    _askSuccess(_testBlendRadiusOnSecondCommand.__name__)
 
 # Tests the blending of two lin motions.
 # Both lin goals have different orientations (difference in euler angle A, B

--- a/pilz_trajectory_generation/CHANGELOG.rst
+++ b/pilz_trajectory_generation/CHANGELOG.rst
@@ -4,6 +4,7 @@ Changelog for package pilz_trajectory_generation
 
 Forthcoming
 -----------
+* fixed an error that led to trajectories not strictly increasing in time
 * update dependencies of trajectory_generation
 * fix CIRC path generator and increase test coverage
 * adopt strictest limits in ptp planner (refactor JointLimitsContainer and TrajectoryGeneratorPTP)

--- a/pilz_trajectory_generation/src/plan_components_builder.cpp
+++ b/pilz_trajectory_generation/src/plan_components_builder.cpp
@@ -76,7 +76,7 @@ void PlanComponentsBuilder::blend(robot_trajectory::RobotTrajectoryPtr other,
   }
 
   // Append the new trajectory elements
-  traj_cont_.back()->append(*blend_response.first_trajectory, 0.0);
+  appendWithStrictTimeIncrease(*(traj_cont_.back()),*blend_response.first_trajectory);
   traj_cont_.back()->append(*blend_response.blend_trajectory, 0.0);
   // Store the last new trajectory element for future processing
   traj_tail_ = blend_response.second_trajectory; // first for next blending segment

--- a/pilz_trajectory_generation/test/integrationtest_command_list_manager.cpp
+++ b/pilz_trajectory_generation/test/integrationtest_command_list_manager.cpp
@@ -203,6 +203,40 @@ TEST_F(IntegrationTestCommandListManager, concatThreeSegments)
   EXPECT_LT(fabs((t1_2_3-t1-t2-t3)), 0.4);
 }
 
+
+/**
+ * @brief Tests if times are strictly increasing with selective blending
+ *
+ * Test Sequence:
+ *    1. Generate request with three trajectories where only the first has a blend radius.
+ *    1. Generate request with three trajectories where only the second has a blend radius.
+ *
+ * Expected Results:
+ *    1. Generation of concatenated trajectory is successful.
+ *       All time steps of resulting trajectory are strictly increasing in time
+ *    2. Generation of concatenated trajectory is successful.
+ *       All time steps of resulting trajectory are strictly increasing in time
+ */
+TEST_F(IntegrationTestCommandListManager, concatSegmentsSelectiveBlending)
+{
+  Sequence seq {data_loader_->getSequence("ComplexSequence")};
+  ASSERT_GE(seq.size(), 3u);
+  seq.erase(3, seq.size());
+  seq.setAllBlendRadiiToZero();
+  seq.setBlendRadii(0, 0.1);
+  RobotTrajCont res1 {manager_->solve(scene_, seq.toRequest())};
+  EXPECT_EQ(res1.size(), 1);
+  EXPECT_GT(res1.front()->getWayPointCount(), 0u);
+  EXPECT_TRUE(hasStrictlyIncreasingTime(res1.front())) << "Time steps not strictly positively increasing";
+
+  seq.setAllBlendRadiiToZero();
+  seq.setBlendRadii(1, 0.1);
+  RobotTrajCont res2 {manager_->solve(scene_, seq.toRequest())};
+  EXPECT_EQ(res2.size(), 1);
+  EXPECT_GT(res2.front()->getWayPointCount(), 0u);
+  EXPECT_TRUE(hasStrictlyIncreasingTime(res2.front())) << "Time steps not strictly positively increasing";
+}
+
 /**
  * @brief Tests the concatenation of two ptp commands
  *

--- a/pilz_trajectory_generation/test/integrationtest_command_list_manager.cpp
+++ b/pilz_trajectory_generation/test/integrationtest_command_list_manager.cpp
@@ -223,14 +223,14 @@ TEST_F(IntegrationTestCommandListManager, concatSegmentsSelectiveBlending)
   ASSERT_GE(seq.size(), 3u);
   seq.erase(3, seq.size());
   seq.setAllBlendRadiiToZero();
-  seq.setBlendRadii(0, 0.1);
+  seq.setBlendRadius(0, 0.1);
   RobotTrajCont res1 {manager_->solve(scene_, seq.toRequest())};
   EXPECT_EQ(res1.size(), 1);
   EXPECT_GT(res1.front()->getWayPointCount(), 0u);
   EXPECT_TRUE(hasStrictlyIncreasingTime(res1.front())) << "Time steps not strictly positively increasing";
 
   seq.setAllBlendRadiiToZero();
-  seq.setBlendRadii(1, 0.1);
+  seq.setBlendRadius(1, 0.1);
   RobotTrajCont res2 {manager_->solve(scene_, seq.toRequest())};
   EXPECT_EQ(res2.size(), 1);
   EXPECT_GT(res2.front()->getWayPointCount(), 0u);
@@ -407,7 +407,7 @@ TEST_F(IntegrationTestCommandListManager, blendingRadiusNegative)
 {
   Sequence seq {data_loader_->getSequence("SimpleSequence")};
   ASSERT_GE(seq.size(), 2u);
-  seq.setBlendRadii(0,-0.3);
+  seq.setBlendRadius(0,-0.3);
   EXPECT_THROW(manager_->solve(scene_, seq.toRequest()), NegativeBlendRadiusException);
 }
 
@@ -425,7 +425,7 @@ TEST_F(IntegrationTestCommandListManager, lastBlendingRadiusNonZero)
 {
   Sequence seq {data_loader_->getSequence("SimpleSequence")};
   ASSERT_EQ(seq.size(), 2u);
-  seq.setBlendRadii(1, 0.03);
+  seq.setBlendRadius(1, 0.03);
   EXPECT_THROW(manager_->solve(scene_, seq.toRequest()), LastBlendRadiusNotZeroException);
 }
 
@@ -444,7 +444,7 @@ TEST_F(IntegrationTestCommandListManager, blendRadiusGreaterThanSegment)
 {
   Sequence seq {data_loader_->getSequence("SimpleSequence")};
   ASSERT_GE(seq.size(), 2u);
-  seq.setBlendRadii(0, 42.0);
+  seq.setBlendRadius(0, 42.0);
   EXPECT_THROW(manager_->solve(scene_, seq.toRequest()), BlendingFailedException);
 }
 
@@ -478,7 +478,7 @@ TEST_F(IntegrationTestCommandListManager, blendingRadiusOverlapping)
   tf2::fromMsg(circ.getGoalConfiguration().getPose(), p2);
   auto distance = (p2.translation()-p1.translation()).norm();
 
-  seq.setBlendRadii(1, (distance - seq.getBlendRadius(0) + 0.01) ); // overlapping radii
+  seq.setBlendRadius(1, (distance - seq.getBlendRadius(0) + 0.01) ); // overlapping radii
   EXPECT_THROW(manager_->solve(scene_, seq.toRequest()), OverlappingBlendRadiiException);
 }
 
@@ -579,7 +579,7 @@ TEST_F(IntegrationTestCommandListManager, TestGripperCmdBlending)
   ASSERT_TRUE(seq.cmdIsOfType<Gripper>(1));
 
   // Ensure that blending is requested for gripper commands.
-  seq.setBlendRadii(0, 1.0);
+  seq.setBlendRadius(0, 1.0);
   RobotTrajCont res_vec {manager_->solve(scene_, seq.toRequest())};
   EXPECT_EQ(res_vec.size(), 1);
 }

--- a/pilz_trajectory_generation/test/integrationtest_sequence_action_capability.cpp
+++ b/pilz_trajectory_generation/test/integrationtest_sequence_action_capability.cpp
@@ -184,7 +184,7 @@ TEST_F(IntegrationTestSequenceAction, TestDifferingGroupNames)
 TEST_F(IntegrationTestSequenceAction, TestNegativeBlendRadius)
 {
   Sequence seq {data_loader_->getSequence("ComplexSequence")};
-  seq.setBlendRadii(0, -1.0);
+  seq.setBlendRadius(0, -1.0);
 
   pilz_msgs::MoveGroupSequenceGoal seq_goal;
   seq_goal.request = seq.toRequest();
@@ -213,7 +213,7 @@ TEST_F(IntegrationTestSequenceAction, TestNegativeBlendRadius)
 TEST_F(IntegrationTestSequenceAction, TestOverlappingBlendRadii)
 {
   Sequence seq {data_loader_->getSequence("ComplexSequence")};
-  seq.setBlendRadii(0, 10*seq.getBlendRadius(0));
+  seq.setBlendRadius(0, 10*seq.getBlendRadius(0));
 
   pilz_msgs::MoveGroupSequenceGoal seq_goal;
   seq_goal.request = seq.toRequest();
@@ -243,7 +243,7 @@ TEST_F(IntegrationTestSequenceAction, TestTooLargeBlendRadii)
 {
   Sequence seq {data_loader_->getSequence("ComplexSequence")};
   seq.erase(2, seq.size());
-  seq.setBlendRadii(0, 10*seq.getBlendRadius(seq.size()-2));
+  seq.setBlendRadius(0, 10*seq.getBlendRadius(seq.size()-2));
 
   pilz_msgs::MoveGroupSequenceGoal seq_goal;
   seq_goal.request = seq.toRequest();
@@ -492,7 +492,7 @@ TEST_F(IntegrationTestSequenceAction, TestIgnoreRobotStateForPlanOnly)
 TEST_F(IntegrationTestSequenceAction, TestNegativeBlendRadiusForPlanOnly)
 {
   Sequence seq {data_loader_->getSequence("ComplexSequence")};
-  seq.setBlendRadii(0, -1.0);
+  seq.setBlendRadius(0, -1.0);
 
   pilz_msgs::MoveGroupSequenceGoal seq_goal;
   seq_goal.request = seq.toRequest();

--- a/pilz_trajectory_generation/test/integrationtest_sequence_service_capability.cpp
+++ b/pilz_trajectory_generation/test/integrationtest_sequence_service_capability.cpp
@@ -144,7 +144,7 @@ TEST_F(IntegrationTestSequenceService, TestDifferingGroupNames)
 TEST_F(IntegrationTestSequenceService, TestNegativeBlendRadius)
 {
   Sequence seq {data_loader_->getSequence("ComplexSequence")};
-  seq.setBlendRadii(0, -1.0);
+  seq.setBlendRadius(0, -1.0);
 
   pilz_msgs::GetMotionSequence srv;
   srv.request.commands = seq.toRequest();
@@ -169,7 +169,7 @@ TEST_F(IntegrationTestSequenceService, TestNegativeBlendRadius)
 TEST_F(IntegrationTestSequenceService, TestOverlappingBlendRadii)
 {
   Sequence seq {data_loader_->getSequence("ComplexSequence")};
-  seq.setBlendRadii(0, 10*seq.getBlendRadius(0));
+  seq.setBlendRadius(0, 10*seq.getBlendRadius(0));
 
   pilz_msgs::GetMotionSequence srv;
   srv.request.commands = seq.toRequest();
@@ -195,7 +195,7 @@ TEST_F(IntegrationTestSequenceService, TestTooLargeBlendRadii)
 {
   Sequence seq {data_loader_->getSequence("ComplexSequence")};
   seq.erase(2, seq.size());
-  seq.setBlendRadii(0, 10*seq.getBlendRadius(seq.size()-2));
+  seq.setBlendRadius(0, 10*seq.getBlendRadius(seq.size()-2));
 
   pilz_msgs::GetMotionSequence srv;
   srv.request.commands = seq.toRequest();

